exercise-book/src/cpp-interop.md

```diff
@@ -0,0 +1,290 @@
+# Cpp Interop Exercise
+
+## After completing this exercise you will be able to
+
+- use `autocxx` to deal with C++ data types from Rust in a header-only library
+- write your own thin wrappers for C++ functions with default parameters
+- be able to call template C++ functions from Rust with `autocxx`
+
+
+## Prerequisites
+
+- FFI
+- Some knowledge of C/C++
+
+This exercise does not require writing vast amounts of Rust code, but does require understanding how different parts of the build system in Rust and C++ interact to get the project going.
+
+## Tasks
+
+1. Use `autocxx` to develop bindings to a `rapidcsv.h` and print out the dates in `example.csv`, from within Rust as well as the `RowCount`.
+
+You should get:
+
+```console
+6
+2017-02-24
+2017-02-23
+2017-02-22
+2017-02-21
+2017-02-17
+```
+
+Because there are 5 rows of data and a header row.
+
+2. Use `autocxx` to read `weather.csv` and report average temperature in June:
+
+```console
+1.1999999999999995
+```
+
+A full solution is available at `rust-exercises/exercises-solutions/cpp-interop`.
+
+## Knowledge
+
+We will learn to use `autocxx` to interface Rust and C++, a usually onerous affair.
+
+Modern C++ dependencies can have exceptions, their own heap, variadic templates, generic parameters and all sorts of functionality that is hard to interface with.
+
+`autocxx` helps us automate this bridge, up to a point, and we'll see how to overcome its limitations and understand where they come from.
+
+To get started, we've provided a header-only C++ library called [rapidcsv](https://github.com/d99kris/rapidcsv) inside `rust-exercises/exercise-templates/cpp-interop` that can load and process CSV files.
+
+`autocxx` works by inspecting the declared data structures  in `rapidcsv.h` and generating the necessary methods and types from the Rust side.
+
+The types and methods for `rapidcsv::Document`, the main data type that holds a CSV, can be generated with
+
+```rust [], ignore
+autocxx::include_cpp! {
+    #include "wrapper.h"
+    generate!("rapidcsv::Document")
+    safety!(unsafe_ffi)
+}
+```
+
+### build.rs
+
+In order to setup `autocxx`, one needs a `src/build.rs` script that will regenerate the necessary bindings on each `cargo build` invocation if `main.rs` changed.
+
+We've taken a very basic script from the `autocxx` repo and dropped it into `src/build.rs`.
+
+`build.rs` files are scripts that are run before the rest of your program, and we'll use it to link with 3rd party code dependencies.
+
+### CxxStrings
+
+We'll define `CxxStrings` with the use of the handy macro
+
+```rust [], ignore
+let_cxx_string!(file_name = "example.csv");
+```
+
+which comes with `cxx`. This can become a very painful constructor otherwise since we'd have to turn our Rust strings into `CxxStrings` for opening a file with a C++ method, so this is very handy.
```

Comment:
```suggestion
which comes with `cxx`. This can become a very painful constructor otherwise since we'd have to turn our Rust strings into `CxxStrings` for opening a file with a C++ method, so this is very handy.

Note that converting these to `&str`s uses at most a fat pointer as overhead, but incurs no further memory allocation on the string object itself.
```